### ISSUE TITLE
CANNOT using someonewhocares.org cause ENCODING or DOWNLOAD issue

### ISSUE THREAD

**Issue by FadeMind** [StevenBlack/hosts#514]
When I add `someonewhocares.org` directory to `data` dir, then I cannot update hosts file due error:
Python2 - cannot finish download hosts file from `someonewhocares.org`
Python3 - finish download file, but ENCODING error made:
```
[fademind@manjaro hsts.test]$ python3 updateHostsFile.py -a -c
Updating source data/add.Risk from https://raw.githubusercontent.com/FadeMind/hosts.extras/master/add.Risk/hosts
Traceback (most recent call last):
  File "updateHostsFile.py", line 1416, in <module>
    main()
  File "updateHostsFile.py", line 167, in main
    update_all_sources(source_data_filename, settings["hostfilename"])
  File "updateHostsFile.py", line 603, in update_all_sources
    update_data = json.load(update_file)
  File "/usr/lib/python3.6/json/__init__.py", line 296, in load
    return loads(fp.read(),
  File "/usr/lib/python3.6/encodings/ascii.py", line 26, in decode
    return codecs.ascii_decode(input, self.errors)[0]
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 27: ordinal not in range(128)
```

When I delete `someonewhocares.org` dir from `data` dir generate hosts file went FINE:

```
[fademind@manjaro ~]$ uhf
Updating source data/add.Risk from https://raw.githubusercontent.com/FadeMind/hosts.extras/master/add.Risk/hosts
Updating source data/CoinBlockerLists from https://raw.githubusercontent.com/ZeroDot1/CoinBlockerLists/master/hosts_browser
Updating source data/adaway.org from https://raw.githubusercontent.com/AdAway/adaway.github.io/master/hosts.txt
Updating source data/tyzbit from https://raw.githubusercontent.com/tyzbit/hosts/master/data/tyzbit/hosts
Updating source data/StevenBlack from https://raw.githubusercontent.com/StevenBlack/hosts/master/data/StevenBlack/hosts
Updating source data/add.2o7Net from https://raw.githubusercontent.com/FadeMind/hosts.extras/master/add.2o7Net/hosts
Updating source data/UncheckyAds from https://raw.githubusercontent.com/FadeMind/hosts.extras/master/UncheckyAds/hosts
Updating source data/mvps.org from http://winhelp2002.mvps.org/hosts.txt
Updating source data/hpHosts-ATS from https://hosts-file.net/ad_servers.txt
Updating source data/add.Dead from https://raw.githubusercontent.com/FadeMind/hosts.extras/master/add.Dead/hosts
Updating source data/hpHosts-EMD from https://hosts-file.net/emd.txt
Updating source data/yoyo.org from https://pgl.yoyo.org/adservers/serverlist.php?hostformat=hosts&mimetype=plaintext&useip=0.0.0.0
Updating source data/hpHosts-MMT from https://hosts-file.net/mmt.txt
Updating source data/KADhosts from https://raw.githubusercontent.com/azet12/KADhosts/master/KADhosts.txt
Updating source data/Badd-Boyz-Hosts from https://raw.githubusercontent.com/mitchellkrogza/Badd-Boyz-Hosts/master/hosts
Updating source data/malwaredomainlist.com from http://www.malwaredomainlist.com/hostslist/hosts.txt
Updating source data/Spotify-Ad-free from https://raw.githubusercontent.com/CHEF-KOCH/Spotify-Ad-free/master/Spotifynulled.txt
Updating source data/add.Spam from https://raw.githubusercontent.com/FadeMind/hosts.extras/master/add.Spam/hosts
Updating source extensions/fakenews from https://raw.githubusercontent.com/marktron/fakenews/master/fakenews
Updating source extensions/social from https://raw.githubusercontent.com/Sinfonietta/hostfiles/master/social-hosts
Updating source extensions/porn/clefspeare13 from https://raw.githubusercontent.com/Clefspeare13/pornhosts/master/0.0.0.0/hosts
Updating source extensions/porn/sinfonietta-snuff from https://raw.githubusercontent.com/Sinfonietta/hostfiles/master/snuff-hosts
Updating source extensions/porn/sinfonietta from https://raw.githubusercontent.com/Sinfonietta/hostfiles/master/pornography-hosts
Updating source extensions/gambling from https://raw.githubusercontent.com/Sinfonietta/hostfiles/master/gambling-hosts
==>0.0.0.0 12.170.116.68<==
==>0.0.0.0 164.109.51.67<==
==>0.0.0.0 5.100.249.215<==
==>0.0.0.0 74.112.173.77<==
==>0.0.0.0 91.205.157.38<==
==>0.0.0.0 91.212.132.230<==
Success! The hosts file has been saved in folder 
It contains 246,768 unique entries.
Moving the file requires administrative privileges. You might need to enter your password.
Flushing the DNS cache to utilize new hosts file...
Flushing the DNS cache requires administrative privileges. You might need to enter your password.
Flushing the DNS cache by restarting NetworkManager.service succeeded
Flushing the DNS cache by restarting dnsmasq.service succeeded
Flushing the DNS cache by restarting NetworkManager.service succeeded
Flushing the DNS cache by restarting dnsmasq.service succeeded
```

**Comment by StevenBlack** [comment]
Is this still a problem Tomasz @FadeMind?  I've generated several times in the past few days and not seen this issue....

**Comment by FadeMind** [comment]
This isssue still exists. Thats why I create this issue task here, regards.

**Comment by notDavid** [comment]
It's a duplicate of #494 ?

**Comment by StevenBlack** [comment]
@notDavid yep, it's a dupe.

I think I need someone who gets this error to submit a PR.  This is fine on MacOS.

**Comment by FadeMind** [comment]
@StevenBlack Issue ASCII encoding was reported here : https://github.com/StevenBlack/hosts/issues/420

**Comment by notDavid** [comment]
@StevenBlack 
>  This is fine on MacOS.

I have this issue on macOS. But not if i use the default configuration (of the hosts repo).

This is how i can reproduce it:

1. Delete the hosts folder
2. run `git clone https://github.com/StevenBlack/hosts.git`
3. run `python3 updateHostsFile.py --auto --backup --replace --flush-dns-cache`
	This finishes succesfully
4. create the folder `hosts/extensions/hpHosts/psh`
5. inside the psh folder create `update.json`:
```
{
    "name": "hpHosts - phishing",
    "description": "This file contains phishing sites listed in the hpHosts database",
    "homeurl": "https://hosts-file.net/?s=Download",
    "frequency": "occasional",
    "issues": "",
    "url": "https://hosts-file.net/psh.txt",
    "license": ""
}
```
6. run `python3 updateHostsFile.py --auto --backup --replace --flush-dns-cache  --extensions hpHosts`
7. You will now see an error: 
```
Problem getting file:  https://hosts-file.net/psh.txt
Error in updating source:  https://hosts-file.net/psh.txt
```
8. Go to the folder `hosts/extensions/hpHosts/psh` in a terminal and run this to manually download the file:
```
wget https://hosts-file.net/psh.txt
mv psh.txt hosts
```
9. run `python3 updateHostsFile.py --auto --backup --replace --flush-dns-cache  --extensions hpHosts`
10. You will now see this error:
```
Traceback (most recent call last):
  File "updateHostsFile.py", line 1416, in <module>
    main()
  File "updateHostsFile.py", line 186, in main
    merge_file = create_initial_file()
  File "updateHostsFile.py", line 653, in create_initial_file
    write_data(merge_file, curFile.read())
  File "/usr/local/Cellar/python3/3.6.4_2/Frameworks/Python.framework/Versions/3.6/lib/python3.6/codecs.py", line 321, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xf1 in position 4812469: invalid continuation byte
```

**Comment by FadeMind** [comment]
@notDavid  in somewherewhocares is this:
```
0.0.0.0 ɢoogle.com
```
at 7675 line.

**Comment by cbuijs** [comment]
@FadeMind 

Looks like this in **ASCII** (line 7675):

0.0.0.0 É¢oogle.com

There is another one (line 9765):

0.0.0.0 secret.É¢oogle.com

This is not compliant to RFC as it should be punnycoded (xn--), the "g" is unicoded (this is a good example of a [IDN homograph](https://en.wikipedia.org/wiki/IDN_homograph_attack) attack).

**Comment by funilrys** [comment]
Such issue could be bypassed with `.encode("idna")` but the protocol of @notDavid is more intriguing :thinking: as that would mean that the `hosts` file is not `utf-8` encoded ...

**Comment by funilrys** [comment]
I think (from my last comment) that we can put an explanation on that issue ... I don't if it has been already done but :

We (and Steven @StevenBlack) assumes that the format is always `utf-8` which is far from true as we can see every time someone got `Error in updating source: xxx`

Actually, this is one of the world all-time issues as we can not detect the format of a file or byte.
I also think that using a library like [BeautifulSoup](https://pypi.python.org/pypi/beautifulsoup4) or [chardet](https://pypi.python.org/pypi/chardet) may help but would it be without performance issues Steven @StevenBlack ?

**Comment by StevenBlack** [comment]
@funilrys don't really mind a performance price here.  I just need this problem to go away 😄

**Comment by funilrys** [comment]
Okay Steven @StevenBlack I'm going to try one of that library to see if it fixes @notDavid protocol then.

**Comment by funilrys** [comment]
@StevenBlack as those libraries are not built-in Python should I include a `requirement.txt` ? :thinking:

**Comment by cbuijs** [comment]
@StevenBlack @funilrys

Trying to understand... The file is already UTF-8 (see below) but the input routine mistakenly takes it as ASCII and trips on decoding. Right?

Your are now saying to fix this, you want to make sure it will be UTF-8 by converting/mangling it after/during download? As I see it, the detection that it is ASCII is wrong, not that the file is not UTF-8.

Just trying to understand and help in any possible capacity ;-).

```
# wget -4q "http://someonewhocares.org/hosts/zero/hosts" -O - | file -
/dev/stdin: UTF-8 Unicode text
```

**Comment by cbuijs** [comment]
@funilrys You might check [Python-Magic](https://github.com/ahupp/python-magic) as well.

**Comment by FadeMind** [comment]
@StevenBlack This is not my code, see: 
https://www.wykop.pl/wpis/30296299/mirki-i-mirableki-z-programowanie-python-problem-j/

> If the contents is encoded with an ASCII based encoding other
than utf-8 (e.g. latin-1), then an appropriate encoding name must
be specified.

> and what is happening in line 603 of the updateHostsFile.py file, the script expects something different than it gets (ascii vs latin-1), and it crashes.


how about:
```
# File Logic
def create_initial_file():
    """
    Initialize the file in which we merge all host files for later pruning.
    """

    merge_file = tempfile.NamedTemporaryFile()

    def safe_decode(source, codec='utf-8'):
        lines = []
        for line in source:
            try:
                lines.append(line.decode(codec))
            except UnicodeDecodeError:
                print('Dropped line:', line)
        return ''.join(lines)

    # spin the sources for the base file
    for source in recursive_glob(settings["datapath"],
                                 settings["hostfilename"]):
        with open(source) as curFile:
            write_data(merge_file, safe_decode(curFile))

    # spin the sources for extensions to the base file
    for source in settings["extensions"]:
        for filename in recursive_glob(path_join_robust(
                settings["extensionspath"], source), settings["hostfilename"]):
            with open(filename) as curFile:
                write_data(merge_file, safe_decode(curFile))

    if os.path.isfile(settings["blacklistfile"]):
        with open(settings["blacklistfile"]) as curFile:
            write_data(merge_file, safe_decode(curFile))

return merge_file
```

BUT now I have:
```
$ python3 updateHostsFile.py -a 
File "updateHostsFile.py", line 668
return merge_file
^
SyntaxError: 'return' outside function
```

**Comment by funilrys** [comment]
@cbuijs The OP issue is can be reviewed once we did the correct conversion. As it demands (as I said [here](https://github.com/StevenBlack/hosts/issues/514#issuecomment-368959062)) that we read each line in order to `.encode("idna")`

My comments and currently library testing were more about @notDavid protocol as I cannot reproduce OP issues...

Between can you provide a `git diff` @FadeMind cause a code out of nothing is kinda hard to follow ...

**Comment by funilrys** [comment]
Just a question @FadeMind which Terminal Emulator are you using?

**Comment by funilrys** [comment]
Hey Steven @StevenBlack !! 
I wrote something that fixes (at least on my side) @notDavid's protocol but coming to test I don't pass the following can you explain me the attended output? 

https://github.com/StevenBlack/hosts/blob/master/testUpdateHostsFile.py#L1365-L1373 :
```
@mock.patch("updateHostsFile.urlopen",
                side_effect=mock_url_open)
    def test_read_url(self, _):
        url = b"www.google.com"

        expected = "www.google.com"
        actual = get_file_by_url(url)

        self.assertEqual(actual, expected)
```


Thanks in advance for your time.

**Comment by funilrys** [comment]
Please forgot about it I found out by myself. Commit + PR is coming after further local tests on multiple data input.

**Comment by FadeMind** [comment]
I can confirm that PR from @funilrys  https://github.com/StevenBlack/hosts/pull/520 RESOLVED issue. CC @StevenBlack

**Comment by ScriptTiger** [comment]
We have been battling this one a long time. Two other issues were https://github.com/StevenBlack/hosts/issues/465 and https://github.com/StevenBlack/hosts/issues/440. I think @funilrys may have just closed half our tickets :P

**Comment by funilrys** [comment]
We can now close this @StevenBlack :+1:
